File: leetcode/sliding_window/SubarraysDivByK.java
/*
974. Subarray Sums Divisible by K
Medium
Topics
Companies
Given an integer array nums and an integer k, return the number of non-empty subarrays that have a sum divisible by k.

A subarray is a contiguous part of an array.
*/

class SubarraysDivByK {
    private static int subarraysDivByK(int[] nums, int k) {
    	int[] map = new int[k];
    	map[0] = 1;
        int start = 0, currSum = 0, count = 0;

        for(int end = 0; end<nums.length; end++) {
        	currSum = (currSum + nums[end]) % k;
        	if (currSum < 0) currSum += k;
        	count += map[currSum];
        	map[currSum]++;
        }
        return count;
    }

    public static void main(String[] args) {
    	int[] nums = {4,5,0,-2,-3,1};
    	int k = 5;
    	System.out.println(subarraysDivByK(nums, k));
    }
}

File: leetcode/sliding_window/caterpillar_count_subarrays/subarrays_div_by_k.py
"""
974. Subarray Sums Divisible by K
Solved
Medium
Topics
Companies
Given an integer array nums and an integer k, return the number of non-empty subarrays that have a sum divisible by k.

A subarray is a contiguous part of an array.
"""


class SubarraysDivByK:
    def subarraysDivByK(self, nums: List[int], k: int) -> int:
        hmap = defaultdict(int)
        hmap[0] = 1
        currSum = 0
        numSub = 0

        for i in range(len(nums)):
            currSum = (currSum + nums[i]) % k
            if currSum < 0:
                currSum += k

            numSub += hmap[currSum]
            hmap[currSum] += 1

        return numSub
